### PR TITLE
Single nozzle filament change

### PR DESCRIPTION
**Pull request by InsanityAutomation**
Complete singlenozzle tool change including temperature and cooling setting storage per material, prime and retract, and park position for material change and temperature changes to prevent dripping on part.

### CODE REVIEW

File: Marlin/src/lcd/ultralcd.cpp [hunk 1]
```diff
@@ -3432,6 +3432,10 @@ void lcd_quick_feedback(const bool clear_buttons) {
         MENU_ITEM(submenu, MSG_IDEX_MENU, IDEX_menu);
       #endif
 
+      #if ENABLED(SINGLENOZZLE)
+        MENU_ITEM_EDIT(float3, MSG_FILAMENT_SWAP_LENGTH, &filemant_swap_length, 0, 200);
```
Comment (thinkyhead) [inline Marlin/src/lcd/ultralcd.cpp]: “Filemant”?

File: Marlin/src/inc/SanityCheck.h [hunk 2]
```diff
@@ -602,6 +602,12 @@ static_assert(X_MAX_LENGTH >= X_BED_SIZE && Y_MAX_LENGTH >= Y_BED_SIZE,
     #error "EXTRUDERS must be 1 with HEATERS_PARALLEL."
   #endif
 
+  #if ENABLED(SINGLENOZZLE)
+    #ifndef(SINGLENOZZLE_FILAMENTLENGTH)
+      #define SINGLENOZZLE_FILAMENTLENGTH 0.0;
+    #endif
+  #endif
+
```
Comment (thinkyhead) [inline Marlin/src/inc/SanityCheck.h:608]: This belongs in one of the Conditionals files, not here.

Comment (InsanityAutomation) [reply Marlin/src/inc/SanityCheck.h:608]: Moved to conditionals_post

**Comment by tcm0116** [comment]
Traditionaly, this has been handled in the "On tool change" script in the slicer or by using firmware retraction. Since this is a subset of firmware retraction, why not just use that?

**Comment by InsanityAutomation** [comment]
The issue in the slicers comes down to the first layer. There has been an open issue in simplify3d for about 2 years where the first layer expects both extruders to have primed material already so it does not advance the filament correctly. This also allows you to tune and adjust the amounts on the fly if desired.

I considered a separate set that would pass off to Firmware retraction, however I did not want to make the tool change function relying on that.

**Comment by InsanityAutomation** [comment]
I suppose another point is that over time every other multi extruder setup has gotten support for a script-less firmware handled tool change. May as well natively support this one as it becomes more common. Could have said the same for idex auto park mode, which is essentially the default nowadays.

No easy way in any of the slicers I've found to differentiate between a chimera and a cyclops style hotend to determine if you actually need to prime on first use of e1 or not. Scripts put alot in the users hands, and this is a simple method to eliminate it all together.

**Comment by Roxy-3D** [comment]
> where the first layer expects both extruders to have primed material already so it does not advance the filament correctly. 

Yeah...   The big problem is...  Both extruders might have been primed.   But with one of them sitting there hot, some material ooozes out.   So any time there is a tool change, some thing has to happen to get the nozzle fully primed again.   I wish Slic3r would allow a 'Prime Tower' where it would just let you build a 30 mm tower for each extruder with so many perimeters a short distance from the part being printed.   And the user could select how many perimeters to do.   That way...   it would be very clear what was needed to get good, high quality tool changes.

**Comment by InsanityAutomation** [comment]
@Roxy-3d this is only for single nozzle dual extruders like a Bowden y tube. Not physically possible to prime both.

**Comment by Roxy-3D** [comment]
Yes!  Understood...   I was just lamenting about the general case....

**Comment by InsanityAutomation** [comment]
Yeah, overall without firmware support, slicers would need to explicitly know if it's a merging single nozzle hotend or not. The next step from here would be storing and recalling the set temp for the other material if different, so retract, set and stabalize temp, prime, go. But that's a separate item at this point.

**Comment by tcm0116** [comment]
@Roxy-3D - I've been using this approach with Slic3r to make configurable prime towers: https://github.com/prusa3d/Slic3r/issues/718#issuecomment-368222150

@InsanityAutomation - I guess I'd feel better if it was more clearly documented what the use case is, and if the value was configurable.

Firmware retraction has some nice features for dealing with the swap retraction and recovery. Perhaps we could build in a feature for firmware retraction that can handle the filament swap when a tool change occurs if a swap was not commanded by the slicer. This would create a unified approach instead of disparate schemes that all sort of do the same thing.

**Comment by Roxy-3D** [comment]
>  I've been using this approach with Slic3r to make configurable prime towers: prusa3d/Slic3r#718 (comment)

@tcm0116  Thank You!   That is clever!   I'll give that a try!    Maybe the Slic3r people can automate that but for right now, that is very doable!

**Comment by tcm0116** [comment]
I think Slic3r PE will eventually get there, but they're currently super focused on supporting their printers. It'd be great to be able to use the dynamic purge tower with my Diamond-equipped printer. It's been a while since I looked into it, so maybe it's possible now.

**Comment by InsanityAutomation** [comment]
@tcm0116 value defaults to 0 if not set, has a menu to adjust under configuration (hence the comment on live adjustment), and is only enabled for 2 extruders with a single hotend.

Use case is to eliminate user controlled scripts for tool change on single nozzle, as well as avoid slicer condition of initial prime for t1 not utilizing retracted amount. In discussion with 2 different vendors, one who sells machines with a Bowden y splitter, and the other who sells kits, both report having run into issues on the first use of t1 not priming the nozzle.

As the slicer is unaware of the dual extruder configuration, it cannot properly determine if t1 should be expected to be purged or not. Given the distance in play, 80mm on the cr-x from tinymachines, it would either require a massive amount of waste into a prime tower as it gets further into the print or manual intervention on the first use.

From the other end, slicers could add an indication of dual tool or single split tool heads but that requires quite a bit more user setup.

Next step past this will be catching and storing inactive tool temp and fan set commands, and bringing them back from here as well. This would allow an identical gcode file for true dual nozzle (chimera) setups as well as single nozzle (cyclops) setups.

**Comment by thinkyhead** [comment]
Rebased, squashed, cleaned up. I pulled out some debug serial-printing. If you want that restored, go ahead, or let me know and I'll put it back while this is still under testing.
```
git fetch origin
git checkout SingleNozzleFilamentChange
git reset --hard origin/SingleNozzleFilamentChange
```

**Comment by InsanityAutomation** [comment]
I'll probably put some of the debug stuff under the normal debug leveling defines at some point. They're fine there till it's done. Hoping to put the rest of the Todo stuff in tomorrow, but wanted this visible so if I get too far from how you'd want it done I don't cause extra work for anybody by finding out too late :)

**Comment by thinkyhead** [comment]
Just did another rebase / squash and included some bugfixes too. Fetch at your convenience.

**Comment by InsanityAutomation** [comment]
Ok, I'll be poking at it again tonight, took the kid to Legoland for the day!

**Comment by thinkyhead** [comment]
Sure, sure. You went there for the kid.

**Comment by InsanityAutomation** [comment]
That's my story and I'm sticking to it! Lol

**Comment by thinkyhead** [comment]
I see what you're trying to do with fan speeds, but it's not going to work without further changes. The array `fanSpeeds[FAN_COUNT]` stores fan speeds per-fan, not per-extruder. So if you want to have two different fan speeds —one per filament— when there's only a single physical fan, you'll need to completely change the meaning of `fanSpeeds` so that it refers to fan speeds per-tool-index instead of per-fan (when `SINGLENOZZLE` is enabled).

(Or, just always assume that `fanSpeeds[0]` is the nozzle's fan.)

**Comment by thinkyhead** [comment]
Patched with the assumption that `fanSpeeds[0]` is the always nozzle's fan.

**Comment by InsanityAutomation** [comment]
Just took a look, cleaned it up for me, covered a couple items on my todo list as well. In hindsight, I agree on the fan always being 0 for single nozzle. If another fan exists, its for a different purpose.

**Comment by InsanityAutomation** [comment]
@thinkyhead Ok, alot closer - added the park position with a separate enable, configurable z lift (may want to make this generic for all except switching and idex, as they currently use a static 1.0mm lift) and a menu to adjust parameters.

A little more cleanup, add some messages, and if it looks ok ill get the configs i just added moved to every example config

**Comment by thinkyhead** [comment]
I'll do another squash shortly to neaten this up….

**Comment by InsanityAutomation** [comment]
Sounds good. Hoping to have this finished up tonight, and test through tomorrow.  I'll poke some more after dinner.

**Comment by thinkyhead** [comment]
I've got a quick rebase ready. If you've got any more commits to add, please push them first and I'll include them in the rebase.

**Comment by InsanityAutomation** [comment]
Nothing staged atm

**Comment by thinkyhead** [comment]
Ok, added a standard method to set the heating message, since it will likely gain some wider usage in the future.

**Comment by InsanityAutomation** [comment]
Sounds good, thanks for taking the time to chime in on this one. I'll be flashing it to a machine shortly to run some test jobs! One more read through :)

**Comment by thinkyhead** [comment]
Ok, just need to add one more commit to fix the current error.

**Comment by thinkyhead** [comment]
Ok, should be totally perfect now!

**Comment by InsanityAutomation** [comment]
Almost, fan speed doesn't switch correctly at the moment. Last change caught temp changes from the tune menu instead of gcode.

**Comment by InsanityAutomation** [comment]
Ok im sleep deprived lol M106 i had hit T instead of P for the param, worked fine.

**Comment by InsanityAutomation** [comment]
@thinkyhead Everything I ran here looked ok, just had it run a bunch of changes, adjusted temp from lcd and made sure it followed. Looks ready to go from my end too now.

**Comment by InsanityAutomation** [comment]
One last thing actually, the speed sets for retract and prime arent stored to eeprom. Ill get that now.

**Comment by InsanityAutomation** [comment]
Ok, once travis is done all yours @thinkyhead im going to sleep!

**Comment by thinkyhead** [comment]
- Created the `M217` G-code so parameters can be set by G-code (also useful for the EEPROM report and reproducibility).
- Updated the RepRap wiki and Marlin website to document the new `M217` G-code.
- Changed feedrate parameters to mm/m and `uint16_t` for consistency with user-facing feedrates (as used with common movement commands) and to conserve SRAM.
- Changed menu items to use accelerated editing and to take values from 10 to 5400.

**Comment by cdedwards** [comment]
If I start heating up the nozzle while I'm using Tool 1, it will not heat up. I'm also seeing an error message while switching back and forth about the nozzle being too cold.

Simplify starts on Tool 0 always. Whereas my main extruder is actually Tool 1.

**Comment by silentninja1** [comment]
As it stands, it requires heat to change tools, as the system cannot physically change filament positions if cold extrusion prevention is on. We could modify it to simply swap the active extruder as some paths for other tool types do with no_move in this case though.

As for not heating when starting with T1, most likely it is sending a temp to t1, and since t0 is still active, nothing heats up yet. I can poke at it a little and see if I can reproduce that any other way tonight. A serial log while tool changing would help confirm or disprove my initial guess.

**Comment by thinkyhead** [comment]
@InsanityAutomation — Looks like this needs more attention.

**Comment by InsanityAutomation** [comment]
@cdedwards take a look at https://github.com/MarlinFirmware/Marlin/pull/12055 - that will allow you to change active tool even when cold, just prevents movement including retraction. Let me know if this resolves your issues, as it sounds like one is a product of the other.

**Comment by InsanityAutomation** [comment]
@cdedwards also see https://github.com/MarlinFirmware/Marlin/pull/12056 which allows you to continue into the routine without a retraction distance set, but prevents movement aside from extruders if not homed.

**Pull request merged**